kona/src/main.rs

```diff
@@ -0,0 +1,326 @@
+            self.derived_payloads_count,
+            attributes.parent.block_info.number as i64 + 1,
+            attributes.attributes.timestamp,
+            pipeline.origin().unwrap().number,
+        );
+        debug!(target: "kona", "attributes: {:#?}", attributes);
+    }
+
+    async fn advance_l2_cursor(&mut self) -> Result<()> {
+        let next_l2_block = self.l2_block_cursor.block_info.number + 1;
+        match self.l2_provider.l2_block_info_by_number(next_l2_block).await {
+            Ok(bi) => {
+                let tip = self
+                    .chain_provider
+                    .block_info_by_number(bi.l1_origin.number)
+                    .await
+                    .map_err(|e| eyre::eyre!(e))?;
+
+                self.l1_to_l2_block_cursor.insert(tip, bi);
+                self.l2_block_cursor = bi;
+                self.should_advance_l2_block_cursor = false;
+            }
+            Err(err) => {
+                error!(target: "kona", ?err, "Failed to fetch next pending l2 safe head: {}", next_l2_block)
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn handle_exex_notification(
+        &mut self,
+        notification: ExExNotification,
+        pipeline: &mut LocalPipeline,
+    ) -> Result<()> {
+        if let Some(reverted_chain) = notification.reverted_chain() {
+            self.chain_provider.commit(reverted_chain.clone());
+            let l1_block_info = info_from_header(&reverted_chain.tip().block);
+
+            // Insert blobs in the blob provider for the reverted chain
+            for block in reverted_chain.blocks_iter() {
+                let tx_hashes = block.transactions().map(|tx| tx.hash).collect::<Vec<_>>();
+                let blobs = self.ctx.pool().get_all_blobs(tx_hashes)?;
+                let blobs = blobs.into_iter().flat_map(|b| b.1.blobs).collect::<Vec<_>>();
+                self.blob_provider.insert_blobs(block.hash(), blobs);
+            }
+
+            // Find the l2 block cursor associated with the reverted L1 chain tip
+            let Some(l2_cursor) = self.l1_to_l2_block_cursor.get(&l1_block_info) else {
+                bail!("Critical: Failed to get previous cursor for old chain tip");
+            };
+
+            // Reset the pipeline to the previous L2 block cursor
+            self.l2_block_cursor = *l2_cursor;
+            if let Err(err) = pipeline.reset(l2_cursor.block_info, l1_block_info).await {
+                bail!("Critical: Failed to reset pipeline: {:?}", err);
+            }
+        }
+
+        if let Some(committed_chain) = notification.committed_chain() {
+            let tip_number = committed_chain.tip().number; // TODO: ensure this is the right tip
+            self.chain_provider.commit(committed_chain);
+
+            if let Err(err) = self.ctx.events.send(ExExEvent::FinishedHeight(tip_number)) {
+                bail!("Critical: Failed to send ExEx event: {:?}", err);
```

Comment:
I'm wondering if we should actually bail here... Can we just retry sending the notification? Maybe we have a configurable retry limit.

Reply:
Since this is an unbounded channel, it can only error if the receiver half is closed, in which case I don't think retrying will be helpful unless I am missing something. I looked at other exex examples and they also always bail on this case, so it should be fine imo

Reply:
agreed with @merklefruit